module/zfs/dbuf.c

```diff
@@ -4658,6 +4677,11 @@ dbuf_sync_lightweight(dbuf_dirty_record_t *dr, dmu_tx_t *tx)
 	 * See comment in dbuf_write().  This is so that zio->io_bp_orig
 	 * will have the old BP in dbuf_lightweight_done().
 	 */
+	if (dr->dr_dnode->dn_phys->dn_nlevels != 1) {
+		parent_db = dr->dr_parent->dr_dbuf;
+		mutex_enter(&parent_db->db_mtx);
+		rw_enter(&parent_db->db_rwlock, RW_READER);
```

Comment:
Again I don't think we need `db_mtx` locked here, while about `db_rwlock` I am not sure, since I don't think there should be a concurrent updates of the same block pointer any way.